File: framework/lib/value_range_control.dart
part of uxml;

/**
 * Defines base class for controls that have are based on a range of values.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class ValueRangeControl extends Control {
  static ElementDef valuerangecontrolElementDef;
  /** MinValue property definition */
  static PropertyDefinition minValueProperty;
  /** MaxValue property definition */
  static PropertyDefinition  maxValueProperty;
  /** Value property definition */
  static PropertyDefinition valueProperty;
  /** ValueChanged event definition */
  static EventDef valueChangedEvent;

  static EventArgs _valueChangedArg = null;

  ValueRangeControl() : super() {
  }

  /**
   * Sets or return value of slider.
   */
  num get value {
    return getProperty(valueProperty);
  }

  set value(num value) {
    setProperty(valueProperty, value);
  }

  /**
   * Sets or returns minimum value.
   */
  num get minValue {
    return getProperty(minValueProperty);
  }

  set minValue(num value) {
    setProperty(minValueProperty, value);
  }

  /**
   * Sets or returns maximum value.
   */
  num get maxValue {
    return getProperty(maxValueProperty);
  }

  set maxValue(num value) {
    setProperty(maxValueProperty, value);
  }

  static void _valueChangedHandler(Object target,
      Object propDef, Object oldValue, Object newValue) {
    ValueRangeControl control = target;
    control.onValueChanged(newValue);
    if (_valueChangedArg == null) {
      _valueChangedArg = new EventArgs(control); // reuse instance
    }
    _valueChangedArg.source = control;
    control.notifyListeners(valueChangedEvent, _valueChangedArg);
  }

  void onValueChanged(num newValue) {
    invalidateLayout();
  }

  static void _constraintValueChangedHandler(Object target,
      Object propDef, Object oldValue, Object newValue) {
    ValueRangeControl control = target;
    control._applyConstraints();
  }

  void _applyConstraints() {
    if (value < minValue && (!minValue.isNaN)) {
      value = minValue;
    }

    if (maxValue < minValue) {
      maxValue = minValue;
    }

    if (value > maxValue && (!maxValue.isNaN)) {
      value = maxValue;
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => valuerangecontrolElementDef;

  /** Registers component. */
  static void registerValueRangeControl() {
    minValueProperty = ElementRegistry.registerProperty("minvalue",
        PropertyType.NUMBER, PropertyFlags.NONE, _constraintValueChangedHandler,
        0.0);
    maxValueProperty = ElementRegistry.registerProperty("maxvalue",
        PropertyType.NUMBER, PropertyFlags.NONE, _constraintValueChangedHandler,
        100.0);
    valueProperty = ElementRegistry.registerProperty("value",
        PropertyType.NUMBER, PropertyFlags.NONE, _valueChangedHandler, 0.0);
    valueChangedEvent = new EventDef("ValueChanged", Route.DIRECT);
    valuerangecontrolElementDef = ElementRegistry.register("ValueRangeControl",
        Control.controlElementDef, [valueProperty, minValueProperty,
        maxValueProperty], [valueChangedEvent]);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/ValueRangeControl.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.EventArgs;
import com.hello.uxml.tools.framework.events.EventDefinition;
import com.hello.uxml.tools.framework.events.EventManager;

import java.util.EnumSet;

/**
 * Defines base class for controls that are based on a range of values.
 *
 * @author ferhat
 */
public class ValueRangeControl extends Control {

  /** Value Property Definition */
  public static PropertyDefinition valuePropDef = PropertySystem.register("Value",
      Double.class, ValueRangeControl.class,
      new PropertyData(0.0, EnumSet.of(PropertyFlags.None), new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
            ((ValueRangeControl) e.getSource()).onValueChanged(
                ((Double) e.getNewValue()).doubleValue());
        }}));

  /** MinValue Property Definition */
  public static PropertyDefinition minValuePropDef = PropertySystem.register("MinValue",
      Double.class, ValueRangeControl.class,
      new PropertyData(0.0));

  /** MaxValue Property Definition */
  public static PropertyDefinition maxValuePropDef = PropertySystem.register("MaxValue",
      Double.class, ValueRangeControl.class,
      new PropertyData(100.0));

  /** ValueChanged event definition */
  public static EventDefinition valueChangedEvent = EventManager.register(
      "ValueChanged", ValueRangeControl.class, EventArgs.class, ValueRangeControl.class, null);

  private EventArgs eventArgs;

  /**
   * Sets or returns value of slider.
   */
  public void setValue(double value) {
    setProperty(valuePropDef, value);
  }

  public double getValue() {
    return ((Double) getProperty(valuePropDef)).doubleValue();
  }

  /**
   * Sets or returns minValue of slider.
   */
  public void setMinValue(double value) {
    setProperty(minValuePropDef, value);
  }

  public double getMinValue() {
    return ((Double) getProperty(minValuePropDef)).doubleValue();
  }

  /**
   * Sets or returns maxValue of slider.
   */
  public void setMaxValue(double value) {
    setProperty(maxValuePropDef, value);
  }

  public double getMaxValue() {
    return ((Double) getProperty(maxValuePropDef)).doubleValue();
  }

  protected void onValueChanged(double newValue) {
    if (newValue < getMinValue()) {
      setValue(getMinValue());
      return;
    }
    if (newValue > getMaxValue()) {
      setValue(getMaxValue());
      return;
    }
    if (hasListener(valueChangedEvent)) {
      if (eventArgs == null) {
        eventArgs = new EventArgs();
        eventArgs.setSource(this);
        eventArgs.setEvent(valueChangedEvent);
      }
      raiseEvent(eventArgs);
    }
  }
}
